### DISCUSSION TITLE
Does cue support patch strategy ?

### DISCUSSION THREAD

**Discussion by rgaoistheone** [cue-lang/cue#1765] [Q&A]
Just wondering if cue support patch strategy like adding elements to list (head or tail ) , or remove / replace a certain element.
for example what I'm expected is ["c", "a", "b"]
```
test: ["a", "b"] & [
  "c", ...,
]
```

And also curious if we support json patch so that we can do live patch to kubernetes resources

**Comment by wonderflow** [comment]
Good to see this discussion, we have the same request and our solution is to hack with annotations:

https://kubevela.io/docs/platform-engineers/cue/patch-strategy

**Reply by slewiskelly** [reply]
> CUE cannot handle conflicting fields currently

I don't think it is a matter of not being able to handle conflicting fields, rather is it a conscious design decision to not allow overriding any concrete value.

This is mentioned throughout the documentation, specifically:
- [About](https://cuelang.org/docs/about/)
- [Configuration](https://cuelang.org/docs/usecases/configuration/)
- [The Logic of CUE](https://cuelang.org/docs/concepts/logic/)

**Reply by rgaoistheone** [reply]
Hi @wonderflow , I'm following Kubevela community and in fact I have a similar discussion here 
[https://github.com/kubevela/kubevela/issues/4164](url)

**Reply by rgaoistheone** [reply]
> > CUE cannot handle conflicting fields currently
> 
> I don't think it is a matter of not being able to handle conflicting fields, rather is it a conscious design decision to not allow overriding any concrete value.
> 
> This is mentioned throughout the documentation, specifically:
> 
> * [About](https://cuelang.org/docs/about/)
> * [Configuration](https://cuelang.org/docs/usecases/configuration/)
> * [The Logic of CUE](https://cuelang.org/docs/concepts/logic/)

Hi @slewiskelly , what I'm trying to here is add c to the test[0] but don't know how to do that .

**Comment by mvdan** [comment]
Your particular example works fine if you concatenate the lists with `+`; the `&` operator unifies the elements instead.

```
$ cat cue.txtar
exec cue export foo.cue

-- foo.cue --
test: ["a", "b"] + [
  "c", ...,
]
$ testscript -v cue.txtar
> exec cue export foo.cue
[stdout]
{
    "test": [
        "a",
        "b",
        "c"
    ]
}
PASS
```

**Reply by rgaoistheone** [reply]
Thanks for your help, but what I'm trying to do is adding c to the test[0] like
```
test: ["c", "a", "b"]
```
Any suggestions ?

**Reply by rg2011** [reply]
I've used the "function pattern" to achieve something similar:

```
$ cat insert.cue
exec cue export insert.cue

-- insert.cue --
#insert: {
  item: _
  at: int & >= 0
  in: [...]
  result: in[0:at] + [item] + in[at:]
}

test: (#insert & { item: "c", at: 0, in: ["a", "b"] }).result

$ testscript -v insert.txtar
> exec cue export insert.cue
[stdout]
{
    "test": [
        "c",
        "a",
        "b"
    ]
}
PASS
```

**Comment by verdverm** [comment]
I've built some non-standard CUE operators for diff, patch, merge, pick, replace, etc... They are only usable from the scripting layer in `hof/flow`, which is based on `cue/flow` (as is `cue cmd`). Since CUE does not allow modifications (as @slewiskelly pointed out), you have to do this in a layer above CUE and the evaluator where you can order operations. Also there is the recursion aspect and being able to do these things with arbitrary objects.

- https://docs.hofstadter.io/data-flow/schemas/tasks/st/
- https://cuetorials.com/deep-dives/recursion/
